[pruning][core][feature] Implement convert for pruner

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97545

Summary:

This PR implements `BaseSparsifier.convert()`, which performs module swapping. 
The modules and mappings will be merged in a future PR.

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseSparsifier.test_convert`

jcaip committed:
[pruning][core][feature] Implement convert for pruner

[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] Implement convert for pruner"




[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] Implement convert for pruner"


Summary:

This PR implements `BaseSparsifier.convert()`, which is used for module swapping in the pruning flow. The modules and mappings will be merged in a future PR.

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseSparsifier.test_convert`
Reviewers:


[ghstack-poisoned]

jerryzh168 commented on torch/ao/pruning/sparsifier/utils.py:
```diff
@@ -1,16 +1,70 @@
+        sparse_mod = mapping[type_before_parametrizations(mod)]
+        new_mod = sparse_mod.from_dense(mod)
+
+        # Preserve module's pre forward hooks. They'll be called on quantized input
+        for pre_hook_fn in mod._forward_pre_hooks.values():
+            new_mod.register_forward_pre_hook(pre_hook_fn)
+        # Preserve module's post forward hooks except _observer_forward_hook
+        # After convert they'll work with quantized output
+        for hook_fn in mod._forward_hooks.values():
+            if hook_fn is not _observer_forward_hook:
+                new_mod.register_forward_hook(hook_fn)
+
+        # respect device affinity when swapping modules
+        devices = {p.device for p in chain(mod.parameters(), mod.buffers())}
+        assert len(devices) <= 1, (
+            "swap_module only works with cpu or single-device CUDA modules, "
+            "but got devices {}".format(devices)
+        )
+        device = next(iter(devices)) if len(devices) > 0 else None
+        if device:
+            new_mod.to(device)
```
nit: looks like this is copied from eager mode quantization? maybe you can refactor this to a separate helper function?

jcaip replied:
I'll add this to the BE tracker and take another look later. 

I'm not sure atm what the best way to reuse the shared code other than passing a function around for the sparse/quant specific steps. And I don't like that b/c I feel like it makes the code harder to read.

Not sure if you have any opinions on this?

jerryzh168 commented on torch/ao/pruning/sparsifier/utils.py:
```diff
@@ -1,16 +1,70 @@
 
 
-def module_to_fqn(model: nn.Module, module: nn.Module, prefix: str = "") -> Optional[str]:
+def module_contains_param(module, parametrization):
```
nit: typing

jerryzh168 commented on torch/ao/pruning/sparsifier/utils.py:
```diff
@@ -1,16 +1,70 @@
+    return False
+
+
+def swap_module(mod, mapping):
```
nit: typing

jerryzh168 commented on torch/ao/pruning/sparsifier/base_sparsifier.py:
```diff
@@ -288,12 +298,45 @@ def squash_mask(self,
-            "`convert` is not implemented. Please, use "
-            "`torch.ao.utils.convert` instead."
-        )
+    def convert(
```
please add types as well

jerryzh168 reviewed (approved):
LGTM

Commit:
Update on "[pruning][core][feature] Implement convert for pruner"


Summary:

This PR implements `BaseSparsifier.convert()`, which performs module swapping. 
The modules and mappings will be merged in a future PR.

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseSparsifier.test_convert`


[ghstack-poisoned]

Commit:
Update on "[pruning][core][feature] Implement convert for pruner"


Summary:

This PR implements `BaseSparsifier.convert()`, which performs module swapping. 
The modules and mappings will be merged in a future PR.

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseSparsifier.test_convert`


[ghstack-poisoned]

Commit:
Update on "[pruning][core][feature] Implement convert for pruner"


Summary:

This PR implements `BaseSparsifier.convert()`, which performs module swapping. 
The modules and mappings will be merged in a future PR.

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseSparsifier.test_convert`


[ghstack-poisoned]

Commit:
Update on "[pruning][core][feature] Implement convert for pruner"


Summary:

This PR implements `BaseSparsifier.convert()`, which performs module swapping. 
The modules and mappings will be merged in a future PR.

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseSparsifier.test_convert`


[ghstack-poisoned]

Commit:
Update on "[pruning][core][feature] Implement convert for pruner"


Summary:

This PR implements `BaseSparsifier.convert()`, which performs module swapping. 
The modules and mappings will be merged in a future PR.

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseSparsifier.test_convert`


[ghstack-poisoned]

jcaip committed:
Update on "[pruning][core][feature] Implement convert for pruner"


Summary:

This PR implements `BaseSparsifier.convert()`, which performs module swapping. 
The modules and mappings will be merged in a future PR.

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseSparsifier.test_convert`


[ghstack-poisoned]

Commit:
Update on "[pruning][core][feature] Implement convert for pruner"


Summary:

This PR implements `BaseSparsifier.convert()`, which performs module swapping. 
The modules and mappings will be merged in a future PR.

Test Plan:
`python test/test_ao_sparsity.py -- TestBaseSparsifier.test_convert`


[ghstack-poisoned]

jcaip commented:
@pytorchbot merge

jcaip commented:
@pytorchbot merge -f "test failures are trunk failures"
